### ISSUE TITLE
`where!(...)` condition is ignored when applied on a query started via association 

### ISSUE THREAD

**Issue by pjpires** [rails/rails#48366]
### Steps to reproduce

I'm using `.where!(...)` (and a few other `bang!` methods) to make my queries composition process a bit more flexible/generic.

Calling `.where!` on top of an existing "base query" started by calling an association on top of a record (e.g. `post.comments`) returns different results from when the query is done via the model class including a where condition (`Comment.where(post: post)`).

I'm aware that these bang methods are currently undocumented (despite still accessible). In any case I thought I should report it as it appears to be an unexpected behavior.

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
    t.string :comment_type
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end


class BugTest < Minitest::Test
  def setup
    @post = Post.create!
    @comment1 = Comment.create!(comment_type: 'Good')
    @comment2 = Comment.create!(comment_type: 'Good')

    @post.comments << @comment1
    @post.comments << @comment2
    @post.comments << Comment.create!(comment_type: 'Bad')
  end

  def test_query_with_comment_where
    query = Comment.where(post: @post)
    query.where!('lower(comment_type) IN (?)', 'good')

    # works
    assert_equal [@comment1, @comment2], query
  end

  def test_query_through_post_association
    query = @post.comments
    query.where!('lower(comment_type) IN (?)', 'good')

    # fails: expected the 2 good comments only, got all 3 comments
    assert_equal [@comment1, @comment2], query
  end
end

```

### Expected behavior

Both tests should run the same query, applying both conditions: `post_id` and `comment_type` and return the same results.

### Actual behavior

Applying `.where!` on top of `post.comments` ignores the `comment_type` condition.

### System configuration
**Rails version**:
7.0.5
**Ruby version**:
3.2.2

**Comment by nvasilevski** [comment]
Hey! Could you elaborate on the need to use `.where!` and not the publicly available `.where`
Cause if we switch the reproduction script to
```ruby
query.where!('lower(comment_type) IN (?)', 'good')
```
to
```ruby
query = query.where('lower(comment_type) IN (?)', 'good')
```

both cases start to pass which makes it really hard to argue that `.where!` is broken since as you pointed out it is not documented and thus private to Rails so currently it may behave in accordance to Rails internal expectations

**Comment by pjpires** [comment]
Hello @nvasilevski,

Thanks for your reply! First of all, my use case might be very specific.

In an internal project of ours we're extensively using complex AR queries for filtering + paginating data, so we added the concept of a "query object" to retrieve data from our DB. This allows us to decouple querying logic from our controllers and models. Since some queries can take options for filtering, paginating and ordering results we end up having a concrete class to organize that code, that also has a respective test file to cover the different scenarios.

Most of our queries follow the same internal steps, so I created a base class for DRYing up some code - `ApplicationQuery`. All queries then extend from this class, and can (optionally) take arguments to filter and paginate data. They must to implement a few methods for the base class to call when running, like building a base query, applying any necessary filtering and ordering the results. I chose to have a `def filter!(data)` in which each query class uses `where!(...)` rather than having know too much about the base class (e.g., doing something like `@data = @data.where(...)`) (which has both advantages and downsides, obviously).

The bang! versions for other AR methods seem to modify the underlying object directly as one would expect, with the exception of `.where!(...)` in this very particular scenario, when the clause is applied on top of a relation. That's why I decided to open this issue - even if it's private to Rails it could still be broken, that was my main point. In any case, if it turns out to be the expected behavior we can close the issue, just thought I should point it out so it could be checked.
